### ISSUE TITLE
EPT schema specification

### ISSUE THREAD

**Issue by connormanning** [connormanning/entwine#109]
Is there a better way to express the schema?  Currently, a `schema` in the main metadata file looks like this:
```
    "schema": [
        { "name": "X", "type": "int32" },
        { "name": "Y", "type": "int32" },
        { "name": "Z", "type": "int32" },
        { "name": "Intensity", "type": "uint16" },
        { "name": "ReturnNumber", "type": "uint8" },
        { "name": "NumberOfReturns", "type": "uint8" },
        { "name": "ScanDirectionFlag", "type": "uint8" },
        { "name": "EdgeOfFlightLine", "type": "uint8" },
        { "name": "Classification", "type": "uint8" },
        { "name": "ScanAngleRank", "type": "float" },
        { "name": "UserData", "type": "uint8" },
        { "name": "PointSourceId", "type": "uint16" },
        { "name": "GpsTime", "type": "double" },
        { "name": "Red", "type": "uint16" },
        { "name": "Green", "type": "uint16" },
        { "name": "Blue", "type": "uint16" },
        { "name": "OriginId", "type": "uint32" }
    ]
```

@m-schuetz brought up a potential improvement [here](https://github.com/pnext/pnext-io/issues/12#issuecomment-404530881), reproduced below:
> In entwine.json the schema consists of tuples with name and type. What about adding "bytes" in addition to that? The number of bytes is already implied by the type, but having the number of bytes in addition to that could make it easier for others to write partial but robust EPT loaders. I'd imagine that over time, the EPT format may get additional types and older EPT loaders could simply skip over the bytes of attributes of an unkown type.

Pre-EPT, the `type` was split out into fields `type` of `signed/unsigned/floating` and `size` of `1/2/4/8`.  They were combined into the above for simplicity, maybe they should be split out?

PDAL uses only these native C-style types from the current EPT schema specification ([here](https://github.com/connormanning/entwine/blob/master/doc/entwine-point-tile.md#schema)) and it seems very unlikely that PDAL will extend its type system, rather types would be coerced into these generic ones.  @m-schuetz, do you have an example of a dimension type, even if hypothetical, where your scenario would come into play?

One shortcoming with the current schema is that it doesn't match laszip storage exactly, which combines certain types into bitfields which are separated in the EPT schema.  There's a note for this in the EPT schema description, but it's still a little odd.

Personally, I like the simplicity of the current schema despite its limitations.  Flexible schema definitions are inherently more complex than a fixed set of schema types (for example the finite set of LAS point types) - so my preference is to keep the schema flexible but as limited as possible to ease implementation.

Thoughts are welcome.

**Comment by hobu** [comment]
For database storage of point clouds, I made an XML schema that looked like this. Instead of type, I called it interpretation. Sizes were in bytes.

```
  <pc:dimension>
    <pc:position>1</pc:position>
    <pc:size>4</pc:size>
    <pc:description>X coordinate as a long integer. You must use the scale and offset information of the header to determine the double value.</pc:description>
    <pc:name>X</pc:name>
    <pc:interpretation>int32_t</pc:interpretation>
    <pc:scale>0.01</pc:scale>
    <pc:active>true</pc:active>
    <pc:uuid>2ee118d1-119e-4906-99c3-42934203f872</pc:uuid>
    <pc:parent_uuid>00000000-0000-0000-0000-000000000000</pc:parent_uuid>
  </pc:dimension>
```

https://github.com/PDAL/PDAL/blob/master/schemas/16-dim-schema.xml


ESRI is doing like this for I3S:

```
        {
          "key": "8",
          "name": "CLASS_CODE",
          "ordering": [
            "attributeValues"
          ],
          "attributeValues": {
            "valueType": "UInt8",
            "valuesPerElement": 1
          }
        },
        {
          "key": "16",
          "name": "FLAGS",
          "ordering": [
            "attributeValues"
          ],
          "attributeValues": {
            "valueType": "UInt8",
            "valuesPerElement": 1
          }
        },
```

https://tiles.arcgis.com/tiles/8cv2FuXuWSfF0nbL/arcgis/rest/services/AUTZEN_LiDAR/SceneServer

**Comment by connormanning** [comment]
I wonder if adding a `count` key, which would be considered `1` if omitted, would be enough to cover these cases.  It seems like array-like values are the main thing that the current schema is missing.  For example:

```
{
        { "name": "X", "type": "int32" },
        { "name": "Y", "type": "int32" },
        { "name": "Z", "type": "int32" },
        { "name": "ArbitraryUserData", "type": "int8", "count": 8 }
}
```

**Comment by connormanning** [comment]
Spent some time discussing/considering this issue today and will implement the following changes:

1. Change the schema specification to be `name`/`type`/`size` to allow for more flexibility in the case of future types which are currently unknown.  For example a string or array type could then be specified and clients implementing only the current 10 types would skip over them with no issue.  Thanks for the suggestion @m-schuetz.

2. Generic scale/offset per-dimension according to #127.

**Comment by connormanning** [comment]
Closed via #139.
